qa/tasks/cephfs/test_snapshots.py

```diff
@@ -442,3 +444,85 @@ def test_multimds_hardlink(self):
 
         self.mount_a.run_shell(["rmdir", "d1/.snap/s1"])
         self.mount_a.run_shell(["rm", "-rf", "d0", "d1"])
+
+    def delete_dir_and_snaps(self, dir_name, snaps):
+        for sno in range(1, snaps+1, 1):
+            sname = "{dir_name}/.snap/s_{sno}".format(dir_name=dir_name, sno=sno)
+            self.mount_a.run_shell(["rmdir", sname])
+        self.mount_a.run_shell(["rmdir", dir_name])
+
+    def create_dir_and_snaps(self, dir_name, snaps):
+        self.mount_a.run_shell(["mkdir", dir_name])
+        
+        # creating first 100 snapshots must be successful
+        for sno in range(1, snaps+1, 1):
+            sname = "{dir_name}/.snap/s_{sno}".format(dir_name=dir_name, sno=sno)
+            try:
+                self.mount_a.run_shell(["mkdir", sname])
+            except CommandFailedError as e:
+                # failing at the last mkdir is expected
+                if sno == snaps:
+                    log.info("failed while creating snap #{}: {}".format(sno, repr(e)))
+                    sys.exc_clear()
+
+
+    def test_mds_max_snaps_per_dir_1_1(self):
+        """
+        Test the newly introudced option named mds_max_snaps_per_dir
+        Default snaps limit is 100
+        Test if the default number of snapshot directories can be created
+        """
+        self.create_dir_and_snaps("accounts", 100)
+        self.delete_dir_and_snaps("accounts", 100)
+
+    def test_mds_max_snaps_per_dir_1_2(self):
+        """
+        Test the newly introudced option named mds_max_snaps_per_dir
+        Default snaps limit is 100
+        Test if the default number of snapshot directory setting is honored.
+        101st directory creation should fail.
+        """
+        self.create_dir_and_snaps("accounts", 101)
+        self.delete_dir_and_snaps("accounts", 100)
+
+    def test_mds_max_snaps_per_dir_2_1(self):
+        """
+        Test the newly introudced option named mds_max_snaps_per_dir
+        Change the default limit to 50 test the configuration path
+        Reduce the snapshot creation limit and test its validity
+        """
+        self.fs.rank_asok(['config', 'set', 'mds_max_snaps_per_dir', '50'])
```

Comment:
Instead subtract 1 from the default value and use that instead.

Reply:
I'm not seeing the value of these tests for 50/150. We do want to test (quoting from previous comment):


* That increasing the snapshot limit allows a previously blocked snapshot (without restarting the MDS)
* That reducing the snapshot limit below the current number of snapshots works and prevents new snapshots from being created.

Reply:
> I'm not seeing the value of these tests for 50/150. We do want to test (quoting from previous comment):
> 
>     * That increasing the snapshot limit allows a previously blocked snapshot (without restarting the MDS)
> 
>     * That reducing the snapshot limit below the current number of snapshots works and prevents new snapshots from being created.

@batrick I think I interpreted the requirement incorrectly. Did you mean to say that all of these tests should be done in a single test function ?

Reply:
@batrick I think I interpreted the requirement incorrectly. Did you mean to say that all of these tests should be done in a single test function ?

Reply:
No. Separate test methods. What I want to see:

* That increasing the snapshot limit allows a previously blocked snapshot (without restarting the MDS)

make snapshots up to the limit; check the next snapshot fails; increase the limit by one; then check the new snapshot succeeds.

* That reducing the snapshot limit below the current number of snapshots works and prevents new snapshots from being created.

make snapshots up to the limit-1. Reduce the limit to limit-2. Then try to create a snapshot.

Reply:
@batrick thanks for elaborating;
I've modified the test methods; please review;